lib/DB/Schema/Result/UserSet/HWSet.pm

```diff
@@ -16,6 +16,10 @@ sub required_dates ($=) {
 	return DB::Schema::Result::ProblemSet::HWSet::required_dates();
 }
 
+sub optional_fields_in_dates ($=) {
+	return DB::Schema::Result::ProblemSet::HWSet::optional_fields_in_dates();
+}
```

Comment:
This doesn't make any sense, and needs to be rethought.  It doesn't make sense to have optional non-date fields in the dates.

The JSON column structure is going to need to be thought out better.  Right now there are several bad design ideas going on with that.  The original idea was for these to be for non-essential uncommon things.  Not dates that are pretty well mapped out.  Furthermore, this is making the implementation rather convoluted.

Reply:
I moved the `enable_reduced_scoring` into the date structure so checking date validity is self-contained for cases of reduced scoring.   This solved the issue of needing to pass in `enable_reduced_scoring` from the `set_params` field to check the validity.  In order to not hard-code this field into the dates, I added this other method.  

Let me know if you have some better thoughts on this or we can revert back to the way I handled it before. 

Although I'm crazy about having a non-date field inside this, it does make a lot of other things easier to handle now.